src/server/pkg/hashtree/hashtree.go

```diff
@@ -0,0 +1,427 @@
+func (h *HashTree) mergeNode(path string, from Interface) (s int64, err error) {
+	if h.Fs == nil {
+		h.Fs = map[string]*Node{}
+	}
+
+	// Fetch the nodes, and return error if one can't be fetched
+	fromNode, err := from.Get(path)
+	if err != nil && Code(err) != PathNotFound {
+		return 0, err
+	}
+	toNode, err := h.Get(path)
+	if err != nil && Code(err) != PathNotFound {
+		return 0, err
+	}
+
+	// Merge 'fromNode' into 'toNode' ('fromNode' will always be defined, because
+	// mergeNode() traverses the 'from' tree)
+	var sizeDelta int64
+	if fromNode.DirNode != nil {
+		if toNode != nil && toNode.DirNode == nil {
+			return 0, errorf(PathConflict, "node at \"%s\" is a directory in the "+
+				"target HashTree, but not in the source", path)
+		}
+		// Create empty directory in 'to' if none exists
+		if toNode == nil {
+			toNode = &Node{
+				Name:    base(path),
+				Size:    0,
+				DirNode: &DirectoryNode{},
+			}
+			h.Fs[path] = toNode
+		}
+
+		// Merge files in 'from' into 'to' (including adding them if they're new)
+		for _, child := range fromNode.DirNode.Children {
+			if s, err := h.mergeNode(join(path, child), from); err == nil {
+				insertStr(&toNode.DirNode.Children, child)
+				sizeDelta += s
+			} else {
+				return 0, err
+			}
+		}
+	} else if fromNode.FileNode != nil {
+		if toNode != nil && toNode.FileNode == nil {
+			return 0, errorf(PathConflict, "node at \"%s\" is a regular file in the "+
+				"source HashTree.Interface, but not in the target", path)
+		}
+		// Create empty file in 'to' if none exists
+		if toNode == nil {
+			toNode = &Node{
+				Name:     base(path),
+				FileNode: &FileNode{},
+			}
+			h.Fs[path] = toNode
+		}
+
+		// Append new blocks
+		toNode.FileNode.BlockRefs = append(
+			toNode.FileNode.BlockRefs, fromNode.FileNode.BlockRefs...)
+
+		// Compute size growth of node
+		for _, br := range fromNode.FileNode.BlockRefs {
+			sizeDelta += int64(br.Range.Upper - br.Range.Lower)
+		}
+	} else {
+		return 0, errorf(Internal,
+			"malformed node at \"%s\": it's neither a file nor a directory", path)
+	}
+	toNode.Size += sizeDelta
+	h.updateHash(path)
+	return sizeDelta, nil
+}
+
+// Merge merges the HashTrees in 'trees' into 'h'. The result is nil if no
+// errors are encountered while merging any tree, or else a new error e, where:
+// - Code(e) is the error code of the first error encountered
+// - e.Error() contains the error messages of all errors encountered
+func (h *HashTree) Merge(trees []Interface) error {
+	h.init()
+
+	errors := []error{}
+	for _, tree := range trees {
+		if _, err := tree.Get("/"); Code(err) == PathNotFound {
+			continue // No work necessary to merge blank tree
+		}
+		_, err := h.mergeNode("", tree) // Empty string is internal repr of "/"
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
+		msgs := make([]string, len(errors))
+		for i := 0; i < len(errors); i++ {
+			msgs[i] = errors[i].Error()
```

Comment:
Why did you want to return all errors, as opposed to just the first error that you encounter?  Is it so that users can resolve all merge errors at once?  I'm a little worried about the fact that the error string is basically unboundedly long, since eventually the error needs to be displayed to the user on some UI.  Can we hard limit the number of errors we return?  Similar to how when you have a lot of compilation errors, the go compiler only displays a couple and then just says "too many errors".